src/_http.ts

```diff
@@ -40,420 +40,455 @@ export type ChalkHttpHeaders = ChalkHttpHeadersFreeform &
-  expires_in: number;
-  engines?: {
-    [name: string]: string;
-  };
+    access_token: string;
+    token_type: string;
+    primary_environment?: string | null;
+    expires_in: number;
+    engines?: {
+        [name: string]: string;
+    };
 }
 
 export class CredentialsHolder {
-  private credentials: ClientCredentials | null = null;
+    private credentials: ClientCredentials | null = null;
 
-  constructor(
-    private config: ChalkClientConfig,
-    private http: ChalkHTTPService
-  ) {}
-
-  async get() {
-    if (this.credentials == null) {
-      try {
-        return (this.credentials = await this.http.v1_oauth_token({
-          baseUrl: this.config.apiServer,
-          body: {
-            client_id: this.config.clientId,
-            client_secret: this.config.clientSecret,
-            grant_type: "client_credentials",
-          },
-        }));
-      } catch (e) {
-        console.error(e);
-        if (isChalkError(e)) {
-          throw e;
-        } else if (e instanceof Error) {
-          throw new ChalkError(e.message);
-        } else {
-          throw new ChalkError(
-            "Unable to authenticate to Chalk servers. Please check your environment config"
-          );
+    constructor(
+        private config: ChalkClientConfig,
+        private http: ChalkHTTPService
+    ) {
+    }
+
+    async get() {
+        if (this.credentials == null) {
+            try {
+                return (this.credentials = await this.http.v1_oauth_token({
+                    baseUrl: this.config.apiServer,
+                    body: {
+                        client_id: this.config.clientId,
+                        client_secret: this.config.clientSecret,
+                        grant_type: "client_credentials",
+                    },
+                }));
+            } catch (e) {
+                console.error(e);
+                if (isChalkError(e)) {
+                    throw e;
+                } else if (e instanceof Error) {
+                    throw new ChalkError(e.message);
+                } else {
+                    throw new ChalkError(
+                        "Unable to authenticate to Chalk servers. Please check your environment config"
+                    );
+                }
+            }
         }
-      }
+
+        return this.credentials;
     }
 
-    return this.credentials;
-  }
+    clear() {
+        this.credentials = null;
+    }
+}
 
-  clear() {
-    this.credentials = null;
-  }
+async function awaitTimeout(timeout: number): Promise<void> {
+    return new Promise((resolve) => {
+        setTimeout(resolve, timeout);
+    });
+}
+
+async function retryFetch(
+    fetch: CustomFetchClient,
+    url: string,
+    options: RequestInit,
+    retries: number,
+    backoff = 100
+): Promise<Response> {
+    let lastError: Error | undefined;
+    for (let i = 0; i < retries; i++) {
+        try {
+            return await fetch(url, options);
+        } catch (e) {
```

Comment:
might in theory be nice to restrict the set of errors that we're willing to retry here more, but w/e